### PR TITLE
DAOS-5664 control: Add retry/cancel logic to dRPC

### PR DESCRIPTION
**Pull request by mjmac**
Add a centralized implementation that all request types
can use. Also allows for requests to be canceled in the
control plane (more work needed to cancel in the ioserver).

### CODE REVIEW

**Review by tanabarr — APPROVED** [review]

File: src/control/drpc/module_svc.go [hunk 1]
```diff
@@ -126,7 +126,7 @@ func (r *ModuleService) ProcessMessage(session *Session, msgBytes []byte) ([]byt
 	}
 	respBody, err := module.HandleCall(session, method, msg.GetBody())
 	if err != nil {
-		r.log.Errorf("HandleCall for %d:%s failed: %s\n", method.String(), method, err)
+		r.log.Errorf("HandleCall for %s:%s failed: %s\n", module, method, err)
```
Comment (jbrosenz) [inline src/control/drpc/module_svc.go:129]: Glad you fixed this one.

File: src/control/server/drpc.go [hunk 2]
```diff
@@ -27,15 +27,69 @@ import (
...
+	defaultRetryAfter = 250 * time.Millisecond
+)
+
+type daosStatusResp struct {
+	Status int32 `protobuf:"varint,1,opt,name=status,proto3" json:"status,omitempty"`
+}
+
+func (dsr *daosStatusResp) String() string {
+	return ""
+}
```
Comment (kjacque) [inline src/control/server/drpc.go:51]: Out of curiosity, why don't we want to print anything here?

File: src/control/server/drpc.go [hunk 3]
```diff
@@ -164,33 +218,78 @@ func newDrpcCall(method drpc.Method, bodyMessage proto.Message) (*drpc.Call, err
...
+			if !retryable {
+				return
+			}
+
+			retryAfter := rdr.RetryAfter
+			if retryAfter == 0 {
+				retryAfter = defaultRetryAfter
+			}
+
+			log.Infof("method %s: retryable %s; retrying after %s", method, status, retryAfter)
```
Comment (jbrosenz) [inline src/control/server/drpc.go:283]: That's nice that the retryAfter time.Duration stringifier handles the units.  Looks good.

**Review by jbrosenz — APPROVED** [review]

**Review by kjacque — APPROVED** [review]

File: src/control/server/drpc.go (hunk 2 above)
Comment (mjmac) [reply src/control/server/drpc.go:51]: I had to add a String() method to satisfy the proto.Message interface. daosStatusResp is just used to extract the status (down in `makeDrpcCall()`) without needing to know the real response type.

**Comment by mjmac** [comment]
The two failures appear to be due to [DAOS-4943](https://jira.hpdd.intel.com/browse/DAOS-4943).

**Pull request merged**